### PR TITLE
Relax overly strict IoContext checking for ReadableStreams

### PR DESCRIPTION
**Pull request by jasnell**
Previous fix broke ability to create a ReadableStream outside of an IoContext.

### CODE REVIEW

File: src/workerd/api/streams/readable.c++ [hunk 1]
```diff
@@ -51,7 +51,9 @@ void ReaderImpl::detach() {
 jsg::Promise<void> ReaderImpl::cancel(
     jsg::Lock& js,
     jsg::Optional<v8::Local<v8::Value>> maybeReason) {
-  ioContext.requireCurrentOrThrowJs();
+  KJ_IF_MAYBE(ctx, ioContext) {
```
Comment (mikea) [inline src/workerd/api/streams/readable.c++:54]: maybe add `assertSameIoContext`? (used here and below)

File: src/workerd/api/streams/standard.c++ [hunk 2]
```diff
@@ -1455,7 +1453,9 @@ bool ReadableStreamDefaultController::hasBackpressure() {
 }
 
 kj::Maybe<int> ReadableStreamDefaultController::getDesiredSize() {
-  ioContext.requireCurrentOrThrowJs();
+  KJ_IF_MAYBE(ctx, ioContext) {
```
Comment (mikea) [inline src/workerd/api/streams/standard.c++:1456]: ditto, assertSameIoContext?

File: src/workerd/api/streams/writable.c++ [hunk 3]
```diff
@@ -32,7 +32,9 @@ jsg::Ref<WritableStreamDefaultWriter> WritableStreamDefaultWriter::constructor(
 jsg::Promise<void> WritableStreamDefaultWriter::abort(
     jsg::Lock& js,
     jsg::Optional<v8::Local<v8::Value>> reason) {
-  ioContext.requireCurrentOrThrowJs();
+  KJ_IF_MAYBE(ctx, ioContext) {
```
Comment (mikea) [inline src/workerd/api/streams/writable.c++:35]: can you use assertSameIoContext?

**Review by mikea — APPROVED** [review]

**Comment by mikea** [comment]
Mostly nits

**Comment by harrishancock** [comment]
What are the new expected behaviors? Could authors use ReadableStreams created at the global scope to communicate between requests?

**Comment by jasnell** [comment]
Indeed they can. I expanded the tests on the internal PR to demonstrate some of it. They can create ReadableStream at the global scope and if you're using only the js promises then it would across requests. However, if you create a RS within a request it can only be used in that request, and you cannot respond with or pipe from a global scope RS. It's .. a bit complicated.

**Comment by kentonv** [comment]
Is this really the right approach, vs. ensuring that IoOwn is used correctly to wrap all KJ I/O objects?

The inconsistency between a ReadableStream created at the global scope vs. in a request seems odd. If someone was broken because they created ReadableStreams at the global scope, doesn't it seem plausible that someone else will be broken because they create them in a request?

**Comment by jasnell** [comment]
Well, it was an attempt to make it easier to reason about the flow through these streams but yeah, it's going to be too restrictive. I've pulled most of the `requireCurrentOrThrowJs()` guards on the methods back out, keeping them only for the pump to loop which is directly interacting with kj i/o.

File: src/workerd/api/streams/standard.c++ [hunk 4]
```diff
@@ -335,28 +347,35 @@ ReadableImpl<Self>::ReadableImpl(
...
+                                            ioContext.addFunctor(kj::mv(onSuccess)),
+                                            ioContext.addFunctor(kj::mv(onFailure)),
+                                            kj::mv(self));
+  } else {
+    algorithms.starting = maybeRunAlgorithm(js,
+                                            algorithms.start,
+                                            kj::mv(onSuccess),
+                                            kj::mv(onFailure),
+                                            kj::mv(self));
+  }
```
Comment (harrishancock) [inline src/workerd/api/streams/standard.c++:377]: Does `maybeRunAlgorithm()` now contain the same logic, so this branch isn't necessary?

File: src/workerd/api/streams/standard.c++ [hunk 5]
```diff
@@ -3068,15 +3079,39 @@ jsg::Promise<void> WritableStreamJsController::pipeLoop(jsg::Lock& js) {
...
+
+  auto promise = pipeLock.source.read(js);
+
+  if (IoContext::hasCurrent()) {
+    auto& ioContext = IoContext::current();
+    return promise.then(js, ioContext.addFunctor(kj::mv(onSuccess)),
+                            ioContext.addFunctor(kj::mv(onFailure)));
+  } else {
+    return promise.then(js, kj::mv(onSuccess), kj::mv(onFailure));
+  }
```
Comment (harrishancock) [inline src/workerd/api/streams/standard.c++:3110]: This conditional continuation wrapping on promises could be worth factoring out.

**Review by harrishancock — APPROVED** [review]

**Pull request merged**